### DISCUSSION TITLE
Using types-lxml types in external code

### DISCUSSION THREAD

**Discussion by roblframpton** [abelcheung/types-lxml#18] [General]
This library does a great job of typing the results from `lxml` functions.  I can see, for example, that the XML `parse` method returns a type `_ElementTree[_Element]`.  However, I cannot see a way to _use_ these types in my own code.  For example, I'd like to write a method which takes an ElementTree as an argument:

```python
def my_func(element: _ElementTree[_Element]):
  pass
```

but I can't work out how to import these types.  I tried:

```python
from types_lxml import _ElementTree, _Element
```

but this didn't work.  So:

1. Is there a way to import these types to use in our own code?
2. Could this be documented somewhere in the readme?

Thanks!

**Comment by abelcheung** [comment]
> Is there a way to import these types to use in our own code?

First things first, actually it is expected to only import classes from source, not from external stubs directly.

And there are a few ways to migitate the problem:

1. Quote the type with double quotes so python interpreter would treat annotation as strings
   ```python
   from lxml.etree import _ElementTree, _Element
   def my_func(element: "_ElementTree[_Element]"):
     pass
   ```
1. Use `__future__`:
   ```python
   from __future__ import annotations
   from lxml.etree import _ElementTree, _Element
   def my_func(element: _ElementTree[_Element]):
     pass

(2) automatically helps achieving (1) for you without your intervention, but you would need to remember adding `__future__` to the very top of every file where it is used.

> Could this be documented somewhere in the readme?

Yes, the documentation absolutely needs some love. I think I can get something done after clearing my current tasks at hand. I'll leave the report open for now as reminder.

**Comment by roblframpton** [comment]
Thanks a lot for your response, that has solved things for me.

**Comment by abelcheung** [comment]
Doc now available [on wiki](https://github.com/abelcheung/types-lxml/wiki/Using-specialised-class-directly).
